File: problem-solving-js/graph.js
// Graph class to represent the graph
class Graph {
  constructor() {
    this.adjacencyList = new Map();
  }

  addNode(node) {
    this.adjacencyList.set(node, []);
  }

  addEdge(node1, node2) {
    this.adjacencyList.get(node1).push(node2);
    this.adjacencyList.get(node2).push(node1); // For an undirected graph
  }

  // BFS function to find the shortest path between two nodes
  bfs(startNode, targetNode) {
    if (!this.adjacencyList.has(startNode) || !this.adjacencyList.has(targetNode)) {
      return null; // Nodes not found in the graph
    }

    const queue = [startNode];
    const visited = new Set();
    const predecessor = new Map();

    visited.add(startNode);

    while (queue.length > 0) {
      const current = queue.shift();

      if (current === targetNode) {
        return this.reconstructPath(predecessor, startNode, targetNode);
      }

      for (const neighbor of this.adjacencyList.get(current)) {
        if (!visited.has(neighbor)) {
          queue.push(neighbor);
          visited.add(neighbor);
          predecessor.set(neighbor, current);
        }
      }
    }

    return null; // No path found between the nodes
  }

  // Helper function to reconstruct the path from the start node to the target node
  reconstructPath(predecessor, startNode, targetNode) {
    const path = [targetNode];
    let current = targetNode;

    while (current !== startNode) {
      current = predecessor.get(current);
      path.unshift(current);
    }

    return path;
  }
}

// Example usage:
const graph = new Graph();

graph.addNode("A");
graph.addNode("B");
graph.addNode("C");
graph.addNode("D");
graph.addNode("E");

graph.addEdge("A", "B");
graph.addEdge("A", "C");
graph.addEdge("B", "D");
graph.addEdge("C", "E");
graph.addEdge("D", "E");

const shortestPath = graph.bfs("A", "E");
console.log("Shortest Path:", shortestPath);

File: problem-solving-rb/graph.rb

# Graph class to represent the graph
class Graph
  def initialize
    @adjacency_list = Hash.new { |h, k| h[k] = [] }
  end

  def add_node(node)
    @adjacency_list[node]
  end

  def add_edge(node1, node2)
    @adjacency_list[node1] << node2
    @adjacency_list[node2] << node1 # For an undirected graph
  end

  # BFS function to find the shortest path between two nodes
  def bfs(start_node, target_node)
    return nil unless @adjacency_list.key?(start_node) && @adjacency_list.key?(target_node)

    queue = [start_node]
    visited = Set.new
    predecessor = {}

    visited.add(start_node)

    while queue.any?
      current = queue.shift

      return reconstruct_path(predecessor, start_node, target_node) if current == target_node

      @adjacency_list[current].each do |neighbor|
        next if visited.include?(neighbor)

        queue.push(neighbor)
        visited.add(neighbor)
        predecessor[neighbor] = current
      end
    end

    nil # No path found between the nodes
  end

  # Helper function to reconstruct the path from the start node to the target node
  def reconstruct_path(predecessor, start_node, target_node)
    path = [target_node]
    current = target_node

    while current != start_node
      current = predecessor[current]
      path.unshift(current)
    end

    path
  end
end

# Example usage:
graph = Graph.new

graph.add_node("A")
graph.add_node("B")
graph.add_node("C")
graph.add_node("D")
graph.add_node("E")

graph.add_edge("A", "B")
graph.add_edge("A", "C")
graph.add_edge("B", "D")
graph.add_edge("C", "E")
graph.add_edge("D", "E")

shortest_path = graph.bfs("A", "E")
puts "Shortest Path: #{shortest_path.inspect}"
